[GTK][WPE] Add support for memory-mapped GPU buffers

nikolaszimmermann opened the pull request:
#### 49ef3002488aa8b83b2f1975816d4172863f1554
<pre>
[GTK][WPE] Add support for memory-mapped GPU buffers
<a href="https://bugs.webkit.org/show_bug.cgi?id=290800">https://bugs.webkit.org/show_bug.cgi?id=290800</a>

Reviewed by Carlos Garcia Campos.

When using CPU rendering we produce pixel data that has to be uploaded
into a GL texture for composition - usually using glTexSubImage2D().

In order to save precious memory bandwidth on embedded devices, explore an
alternative mechanism, where buffer uploading happens out of the scope of
OpenGL to avoid overhead, such as copying the pixel data into a staging
buffer in Mesa etc.

The idea is to create a GBM buffer object, export it as a dma-buf, wrap
the dma-buf in an EGLImage and use it as backing for a regular OpenGL
texture, using glEGLImageTargetTexture2DOES(). Later on to update the
texture, one can mmap() the dma-buf fd, to gain write access to the
dma-buf. All memory operations have to be fenced using DMA_BUF_SYNC_*
system calls to ensure that the data is properly synchronized between
CPU and GPU (on shared memory systems, that are common in embedded).

Optimize BitmapTexture::updateContents() for the case when we want
to update a linear, mappable texture (backed by a dma-buf). Avoid
using glCopyTexSubImage2D, and use memcpy() instead on the mapped
memory region -- that avoids driver pitfalls, giving more control
about when synchronization happens.

Allocate tile textures always using GL_NEAREST instead of GL_LINEAR
filter -- GL_NEAREST is less expensive than GL_LINEAR, consuming
less bandwidth. However, it can only be used for textures that are
not scaled/filtered/... to avoid quality loss.

Add WEBKIT_SKIA_USE_LINEAR_TILE_TEXTURES environment variable,
only when set to 1 linear textures are used for the tiles. Off by
default, since the benefit is highly platform dependant.

Canonical link: <a href="https://commits.webkit.org/294962@main">https://commits.webkit.org/294962@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a08583846541b08d505b5094945d065bc52d59ba

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/103626 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/23309 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/13629 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/108802 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/54273 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/105666 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/23664 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/31859 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/78747 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/106632 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/18357 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/93483 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/59082 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/18170 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/11530 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/53627 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/87951 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/11589 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/111180 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/30778 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/22696 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/87738 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/31134 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/89683 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/87386 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22247 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/32265 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/9987 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/25128 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/30701 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/36013 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/30501 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/33832 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/32062 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aoikonomopoulos commented on Source/WebCore/platform/graphics/gbm/MemoryMappedGPUBuffer.cpp:
```diff
@@ -0,0 +1,505 @@
+            uint32x4_t v0 = vld1q_u32(srcRow + x);
+            uint32x4_t v1 = vld1q_u32(srcRow + x + 1 * sizeof(uint32_t));
+            uint32x4_t v2 = vld1q_u32(srcRow + x + 2 * sizeof(uint32_t));
+            uint32x4_t v3 = vld1q_u32(srcRow + x + 3 * sizeof(uint32_t));
```
I may be misunderstanding the docs, but here you increase the offset by 4 every time because you're loading 4 `uint32_t` values at a time, right? So the `sizeof(uint32_t)` here (which happens to also be 4) seems misleading?

Secondly, this is a memcpy, right? I would imagine an optimized memcpy belongs in WTF/?

Thirdly, could it be that a call to memcpy might be faster here? The libc version may already be using SIMD instructions, right?

nikolaszimmermann replied:
Thanks @aoikonomopoulos and sorry for the late reply -- the code is no longer relevant, I've omitted it from recent versions -- and you were right on the sizeof comment and the memcpy optimization.

spenap commented on Source/WebCore/platform/graphics/gbm/MemoryMappedGPUBuffer.h:
```diff
@@ -0,0 +1,93 @@
+/*
+ * Copyright (C) 2024 Igalia S.L.
```
```suggestion
 * Copyright (C) 2025 Igalia S.L.
```

I first thought that was a typo and this should be 25, but the implementation file says 2024,2025. 

If the file already existed I guess 2024 is OK, but shouldn't header and implementation be aligned?

nikolaszimmermann replied:
I was not aware that header/impl need to be aligned -- never did that at least :-) I only modify copyright dates in general, if I think I made a significant enough change.

nikolaszimmermann replied:
However, fixed and set to `Copyright (C) 2024, 2025 Igalia S.L.` which is acurate.

spenap commented on Source/WebCore/platform/graphics/gbm/MemoryMappedGPUBuffer.cpp:
```diff
@@ -0,0 +1,505 @@
+constexpr bool dumpSupportedDRMFormatsAndModifiers = false;
+constexpr bool dumpBufferFormatNegotiation = false;
+constexpr bool dumpBufferAllocation = false;
+constexpr bool useDRMModifiers = true;
```
do we not mark static-like variables somehow, e.g. with a leading `k` or `s`?

nikolaszimmermann replied:
Switched to `k` prefix -- good suggestion.

spenap commented on Source/WebCore/platform/graphics/gbm/MemoryMappedGPUBuffer.cpp:
```diff
@@ -0,0 +1,505 @@
+
+    auto deviceNode = manager.mainDeviceNode(WebCore::DRMDeviceManager::NodeType::Primary);
+    if (deviceNode)
+        LOG(Images, "[MemoryMappedGPUBuffer] Allocating dma-bufs from device '%s'", deviceNode->filename().data());
```
```suggestion
    if (!deviceNode) {
        LOG(Images, "[MemoryMappedGPUBuffer] Failed to get device node");
        return nullptr;
    }

    LOG(Images, "[MemoryMappedGPUBuffer] Allocating dma-bufs from device '%s'", deviceNode->filename().data());
```

We should have a static checker that flags the above.

If we consider `deviceNode` worth a null check before dereferencing it to log, then we should not unconditionally dereference it to get the `gbmDevice()`

nikolaszimmermann replied:
Good catch, will adapt to:

```
diff --git a/Source/WebCore/platform/graphics/gbm/MemoryMappedGPUBuffer.cpp b/Source/WebCore/platform/graphics/gbm/MemoryMappedGPUBuffer.cpp
index 6dde788c162c..0e3045818331 100644
--- a/Source/WebCore/platform/graphics/gbm/MemoryMappedGPUBuffer.cpp
+++ b/Source/WebCore/platform/graphics/gbm/MemoryMappedGPUBuffer.cpp
@@ -71,8 +71,12 @@ std::unique_ptr<MemoryMappedGPUBuffer> MemoryMappedGPUBuffer::create(const IntSi
     ASSERT(manager.isInitialized());
 
     auto deviceNode = manager.mainDeviceNode(WebCore::DRMDeviceManager::NodeType::Render);
-    if (deviceNode)
-        LOG(GPUBuffers, "MemoryMappedGPUBuffer::create(), allocating dma-bufs from device '%s'", deviceNode->filename().data());
+    if (!deviceNode) {
+        LOG(GPUBuffers, "MemoryMappedGPUBuffer::create(), failed to get GBM render device node");
+        return nullptr;
+    }
+
+    LOG(GPUBuffers, "MemoryMappedGPUBuffer::create(), allocating dma-bufs from device '%s'", deviceNode->filename().data());
 
     auto* device = deviceNode->gbmDevice();
     if (!device) {
```

spenap commented on Source/WebCore/platform/graphics/gbm/MemoryMappedGPUBuffer.cpp:
```diff
@@ -0,0 +1,505 @@
+                    if (modifier == DRM_FORMAT_MOD_LINEAR)
+                        hasLinearModifier = true;
+                    else if (modifier == DRM_FORMAT_MOD_VIVANTE_SUPER_TILED)
+                        hasVivanteSuperTiledModifier = true;
```
```suggestion
                    if (modifier == DRM_FORMAT_MOD_LINEAR)
                        hasLinearModifier = true;
                    else if (modifier == DRM_FORMAT_MOD_VIVANTE_SUPER_TILED)
                        hasVivanteSuperTiledModifier = true;
                    if (hasLinearModifier && hasVivanteSuperTiledModifier)
                        break
```

I understand we want to find out whether these two modifiers are supported. After we have done so, I think we can break. If we have a preference for either, then we could break earlier, but I didn't want to change the behaviour in my suggestion

nikolaszimmermann replied:
I simplified it, please re-check.

spenap commented on Source/WebCore/platform/graphics/gbm/MemoryMappedGPUBuffer.cpp:
```diff
@@ -0,0 +1,505 @@
+                else
+                    useFormat.modifiers = { };
+            } else
+                useFormat.modifiers = { };
```
```suggestion
            useFormat.modifiers = { };
            if (useDRMModifiers) {
                if (hasVivanteSuperTiledModifier && !shouldPreferDRMLinearModifier())
                    useFormat.modifiers = { DRM_FORMAT_MOD_VIVANTE_SUPER_TILED };
                else if (hasLinearModifier)
                    useFormat.modifiers = { DRM_FORMAT_MOD_LINEAR };
            }
```

nikolaszimmermann replied:
Code has changed, please re-check.

spenap commented on Source/WebCore/platform/graphics/gbm/MemoryMappedGPUBuffer.cpp:
```diff
@@ -0,0 +1,505 @@
+bool MemoryMappedGPUBuffer::allocate(struct gbm_device* device, const GLDisplay::DMABufFormat& bufferFormat)
+{
+    m_modifier = DRM_FORMAT_MOD_INVALID;
+    if (bufferFormat.modifiers.size() > 0)
```
```suggestion
    if (!bufferFormat.modifiers.isEmpty())
```

This is a personal preference, and might even be against "custom", but I tend to use empty checks when we don't care about the size

nikolaszimmermann replied:
That's right, preferred WebKit style too. Thanks.

spenap commented on Source/WebCore/platform/graphics/gbm/MemoryMappedGPUBuffer.cpp:
```diff
@@ -0,0 +1,505 @@
+
+        return true;
+    }
+
```
Shouldn't we log an error in this case and the one above, when we fail to export?

nikolaszimmermann replied:
Agreed, we should be consistent:

```
 int MemoryMappedGPUBuffer::exportGBMBufferObjectAsDMABuf(unsigned planeIndex)
 {
     auto handle = gbm_bo_get_handle_for_plane(m_bo, planeIndex);
-    if (handle.s32 == -1)
+    if (handle.s32 == -1) {
+        LOG_ERROR("MemoryMappedGPUBuffer::exportGBMBufferObjectAsDMABuf(), failed to obtain gbm handle for plane %u", planeIndex);
         return -1;
+    }
 
     int fd = 0;
     int ret = drmPrimeHandleToFD(gbm_device_get_fd(gbm_bo_get_device(m_bo)), handle.u32, DRM_CLOEXEC | DRM_RDWR, &fd);
-    return ret < 0 ? -1 : fd;
+    if (ret < 0) {
+        LOG_ERROR("MemoryMappedGPUBuffer::exportGBMBufferObjectAsDMABuf(), failed to export dma-buf for plane %u", planeIndex);
+        return -1;
+    }
+
+    return fd;
 }
```

spenap commented on Source/WebCore/platform/graphics/gbm/MemoryMappedGPUBuffer.cpp:
```diff
@@ -0,0 +1,505 @@
+    ASSERT(m_mappedData);
+    ASSERT(m_mappedLength > 0);
+    munmap(m_mappedData, m_mappedLength);
+}
```
I would suggest we wrap this into an RAII object, even if it is a private / nested object in this class

nikolaszimmermann replied:
Followed that suggestion, we now have:

```
inline std::unique_ptr<MemoryMappedGPUBuffer::AccessScope> makeGPUBufferReadScope(MemoryMappedGPUBuffer& buffer)
{
    return MemoryMappedGPUBuffer::AccessScope::create(buffer, MemoryMappedGPUBuffer::AccessScope::Mode::Read);
}

inline std::unique_ptr<MemoryMappedGPUBuffer::AccessScope> makeGPUBufferWriteScope(MemoryMappedGPUBuffer& buffer)
{
    return MemoryMappedGPUBuffer::AccessScope::create(buffer, MemoryMappedGPUBuffer::AccessScope::Mode::Write);
}
```

and a helper class

```
    class AccessScope {
        WTF_MAKE_NONCOPYABLE(AccessScope);
        WTF_MAKE_FAST_ALLOCATED();
    public:
        enum class Mode : bool {
            Read,
            Write
        };

        static std::unique_ptr<AccessScope> create(MemoryMappedGPUBuffer&, Mode);

        AccessScope(MemoryMappedGPUBuffer&, Mode);
        ~AccessScope();

    private:
        MemoryMappedGPUBuffer& m_buffer;
        Mode m_mode { Mode::Read };
    };
```

```
MemoryMappedGPUBuffer::AccessScope::AccessScope(MemoryMappedGPUBuffer& buffer, AccessScope::Mode mode)
    : m_buffer(buffer)
    , m_mode(mode)
{   
    ASSERT(m_buffer.isMapped());
    m_buffer.performDMABufSyncSystemCall({ DMABufSyncFlag::Start, m_mode == AccessScope::Mode::Read ? DMABufSyncFlag::Read : DMABufSyncFlag::Write });
}   

MemoryMappedGPUBuffer::AccessScope::~AccessScope()
{   
    m_buffer.performDMABufSyncSystemCall({ DMABufSyncFlag::End, m_mode == AccessScope::Mode::Read ? DMABufSyncFlag::Read : DMABufSyncFlag::Write });
}   
```

AccessScope encapsulates the safe access to a memory mapped region, including explicit synchronization.

spenap commented on Source/WebCore/platform/graphics/gbm/MemoryMappedGPUBuffer.cpp:
```diff
@@ -0,0 +1,505 @@
+        return false;
+    }
+
+    auto dmaBufFD = gbm_bo_get_fd_for_plane(m_bo, 0);
```
I am similarly tempted to suggest a RAII object for the file descriptor. I see that `UnixFileDescriptor` exists, could we use it here?

nikolaszimmermann replied:
Done.

spenap commented on Source/WebCore/platform/graphics/gbm/MemoryMappedGPUBuffer.cpp:
```diff
@@ -0,0 +1,505 @@
+bool MemoryMappedGPUBuffer::exportGBMBufferObjectAsDMABuf()
+{
+    if (auto dmaBufFD = gbm_bo_get_fd_for_plane(m_bo, 0); dmaBufFD >= 0) {
+        m_dmaBufFD = dmaBufFD;
```
same -- we could use the UnixFileDescriptor and then we wouldn't need to close it explicitly in the destructor

nikolaszimmermann replied:
Nowadays using `RefPtr<DMABufBuffer> m_dmaBuf` -- comment is no longer applicable, and you were right of course :-)

spenap commented on Source/WebCore/platform/graphics/gbm/MemoryMappedGPUBuffer.cpp:
```diff
@@ -0,0 +1,505 @@
+
+void MemoryMappedGPUBuffer::updateContents(const void* srcData, const IntRect& targetRect, const IntPoint& sourceOffset, unsigned bytesPerLine)
+{
+    ASSERT(m_mappedData);
```
I guess this would be very rare, but. Do you think we need to make this a release assertion, or otherwise protect the case where `dstPixels` becomes null in line 468?

nikolaszimmermann replied:
No longer applicable.

spenap commented on Source/WebCore/platform/graphics/texmap/BitmapTexturePool.cpp:
```diff
@@ -1,6 +1,6 @@
 /*
  * Copyright (C) 2010 Nokia Corporation and/or its subsidiary(-ies)
- * Copyright (C) 2015 Igalia S.L.
+ * Copyright (C) 2015, 2024 Igalia S.L.
```
```suggestion
 * Copyright (C) 2015, 2025 Igalia S.L.
```

nikolaszimmermann replied:
Fixed.

nikolaszimmermann commented on Source/WebCore/platform/mediastream/gstreamer/RealtimeIncomingAudioSourceGStreamer.cpp:
Adding the new file `MemoryMappedGPUBuffer.cpp` to build changed the order of the unified builds packaging --> fix errors in files, unrelated to this patch, to keep the build intact.

nikolaszimmermann commented on Source/WebCore/platform/graphics/texmap/BitmapTexture.h:
```diff
@@ -90,12 +98,14 @@ class BitmapTexture final : public ThreadSafeRefCounted<BitmapTexture> {
 
     ClipStack& clipStack() { return m_clipStack; }
 
-    void copyFromExternalTexture(GLuint textureID);
```
This change could be moved into a separated PR, to avoid the change of the single call site affected by the removal: `Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedPlatformLayerBufferVideo.cpp`. However, the arguments were wrongly named, and this caused confusion, when trying to add MemoryMappedGPUBuffer awareness to those functions, just avoid those call sites, and make the sizing requirements explicit, you need a _targetRect_ and a _sourceOffset_ to invoke the `copyFromExternalTexture` operation.

spenap commented on Source/WebCore/platform/graphics/gbm/MemoryMappedGPUBuffer.h:
```diff
@@ -0,0 +1,141 @@
+    WTF_MAKE_NONCOPYABLE(MemoryMappedGPUBuffer);
+    WTF_MAKE_FAST_ALLOCATED();
+public:
+    MemoryMappedGPUBuffer(const IntSize&);
```
Since we now have a `create` method, should we make `MemoryMappedGPUBuffer` protected or private? 

There is a trick where you can use: https://playfulprogramming.com/posts/a-forgotten-idiom-revisited-pass-key, so that even if you keep the constructor public, you effectively restrict its usage by requiring the key.

nikolaszimmermann replied:
I dislike the fact too that the `MemoryMappedGPUBuffer` has to be public, just for `makeUnique` to work and you're right the technique of passing a flag is well known, however not used at all in WebKit. IMHO it makes sense to formalize a rule, that constructors should be private of classes constructed via a static creation function.

The alternative, used in recent examples of Apple code, is to avoid `makeUnique`:

```
diff --git a/Source/WebCore/platform/graphics/gbm/MemoryMappedGPUBuffer.cpp b/Source/WebCore/platform/graphics/gbm/MemoryMappedGPUBuffer.cpp
index 33c225c841aa..ecb199d744fa 100644
--- a/Source/WebCore/platform/graphics/gbm/MemoryMappedGPUBuffer.cpp
+++ b/Source/WebCore/platform/graphics/gbm/MemoryMappedGPUBuffer.cpp
@@ -163,18 +163,18 @@ std::unique_ptr<MemoryMappedGPUBuffer> MemoryMappedGPUBuffer::create(const IntSi
         return nullptr;
     }
 
-    auto gbmBuffer = makeUnique<MemoryMappedGPUBuffer>(size);
-    if (!gbmBuffer->allocate(device, bufferFormat.value())) {
+    auto buffer = std::unique_ptr<MemoryMappedGPUBuffer>(new MemoryMappedGPUBuffer(size));
+    if (!buffer->allocate(device, bufferFormat.value())) {
         LOG_ERROR("MemoryMappedGPUBuffer::create(), failed to create GBM buffer of size %dx%d: %s", size.width(), size.height(), safeStrerror(errno).data());
         return nullptr;
     }
 
-    if (!gbmBuffer->createDMABufFromGBMBufferObject()) {
+    if (!buffer->createDMABufFromGBMBufferObject()) {
         LOG_ERROR("MemoryMappedGPUBuffer::create(), failed to create dma-buf from GBM buffer object");
         return nullptr;
     }
 
-    return gbmBuffer;
+    return buffer;
 }
 
 bool MemoryMappedGPUBuffer::allocate(struct gbm_device* device, const GLDisplay::DMABufFormat& bufferFormat)
diff --git a/Source/WebCore/platform/graphics/gbm/MemoryMappedGPUBuffer.h b/Source/WebCore/platform/graphics/gbm/MemoryMappedGPUBuffer.h
index 9be3ab1dc162..ac59e2da547c 100644
--- a/Source/WebCore/platform/graphics/gbm/MemoryMappedGPUBuffer.h
+++ b/Source/WebCore/platform/graphics/gbm/MemoryMappedGPUBuffer.h
@@ -46,7 +46,6 @@ class MemoryMappedGPUBuffer {
     WTF_MAKE_NONCOPYABLE(MemoryMappedGPUBuffer);
     WTF_MAKE_FAST_ALLOCATED();
 public:
-    MemoryMappedGPUBuffer(const IntSize&);
     ~MemoryMappedGPUBuffer();
 
     enum class BufferFlag : uint8_t {
@@ -98,6 +97,8 @@ public:
     bool isLinearBuffer() const;
 
 private:
+    MemoryMappedGPUBuffer(const IntSize&);
+
     enum class DMABufSyncFlag : uint8_t {
         Start = 1 << 0,
         End   = 1 << 1,
```

Removes the needs for any "pass-key" like tricks.

spenap commented on Source/WebCore/platform/graphics/gbm/MemoryMappedGPUBuffer.h:
```diff
@@ -0,0 +1,141 @@
+    EGLImage createEGLImageFromDMABuf();
+
+    // Update the underlying data of the dma-buf, as often as desired.
+    // You need to obtain an AccessScope, fencing the write operation.
```
Is there a way we can make this constraint explicit? Can we take the scope as an argument, and assert some of its properties -- as we would do if we were passing in a lock?

nikolaszimmermann replied:
Good idea - better than documentation 😂 will adapt

spenap commented on Source/WebCore/platform/graphics/gbm/MemoryMappedGPUBuffer.cpp:
```diff
@@ -0,0 +1,444 @@
+    constexpr uint32_t preferredDMABufFormat = DRM_FORMAT_ABGR8888;
+
+    const auto& supportedFormats = PlatformDisplay::sharedDisplay().dmabufFormats();
+    if constexpr (kDumpSupportedDRMFormatsAndModifiers) {
```
I was looking into the loops and thinking about avoiding the impact if not logging, then I checked properly and your approach is very good -- a constexpr check. I like this a lot!

spenap commented on Source/WebCore/platform/graphics/gbm/MemoryMappedGPUBuffer.cpp:
```diff
@@ -0,0 +1,444 @@
+        EGL_LINUX_DRM_FOURCC_EXT, static_cast<EGLAttrib>(m_format)
+    };
+
+#define ADD_PLANE_ATTRIBUTES(planeIndex) { \
```
Could we achieve the same without a macro? Is there another way to produce the `EGL_DMA_BUF_PLANE##planeIndex##....`  keys? I find this a bit intimidating. 

It wouldn't be a straight replacement as this macro can `return false` and if you turn it into a lambda, its output would need handling...

nikolaszimmermann replied:
This is a 1:1 copy from another place that defines the same ugly macro - I have a follow-up cleanup PR in the pipeline, that will unify all call sites using this construct, without using macros. Ok to postpone this?

spenap replied:
yes, thanks! It makes more sense to handle in a follow-up

spenap reviewed (commented):
LGTM. Some minor comments but nothing big. Thanks!

spenap reviewed (commented):
LGTM, thanks!!

carlosgcampos commented on Source/WebCore/platform/graphics/gbm/MemoryMappedGPUBuffer.cpp:
```diff
@@ -0,0 +1,463 @@
+#include "IntRect.h"
+#include "Logging.h"
+#include "PlatformDisplay.h"
+
```
```suggestion
```

carlosgcampos commented on Source/WebCore/platform/graphics/gbm/MemoryMappedGPUBuffer.cpp:
```diff
@@ -0,0 +1,463 @@
+                useFormat.modifiers = { };
+                break;
+            }
+
```
```suggestion
```

carlosgcampos commented on Source/WebCore/platform/graphics/gbm/MemoryMappedGPUBuffer.cpp:
```diff
@@ -0,0 +1,463 @@
+                    hasLinearModifier = true;
+                    break;
+                }
+            }
```
I guess you can use `format.modifiers.contains(DRM_FORMAT_MOD_LINEAR)`

carlosgcampos commented on Source/WebCore/platform/graphics/gbm/MemoryMappedGPUBuffer.cpp:
```diff
@@ -0,0 +1,463 @@
+        }
+
+        if (!useFormat.fourcc)
+            return std::nullopt;
```
I think we could move this before the force linear check

nikolaszimmermann replied:
But we'd also need it here, in case no format matches at all, this is the only exit condition.

carlosgcampos commented on Source/WebCore/platform/graphics/gbm/MemoryMappedGPUBuffer.cpp:
```diff
@@ -0,0 +1,463 @@
+        return nullptr;
+    }
+
+    auto buffer = std::unique_ptr<MemoryMappedGPUBuffer>(new MemoryMappedGPUBuffer(size));
```
makeUnique?

nikolaszimmermann replied:
Avoided on purpose to keep the ctor private.

carlosgcampos commented on Source/WebCore/platform/graphics/skia/SkiaPaintingEngine.cpp:
```diff
@@ -239,7 +239,7 @@ Ref<SkiaRecordingResult> SkiaPaintingEngine::record(const GraphicsLayer& layer,
     ASSERT(useThreadedRendering());
 
-    auto renderingMode = canPerformAcceleratedRendering() ? RenderingMode::Accelerated : RenderingMode::Unaccelerated;
+    auto renderingMode = (m_gpuWorkerPool && canPerformAcceleratedRendering()) ? RenderingMode::Accelerated : RenderingMode::Unaccelerated;
```
Is this related to this patch?

nikolaszimmermann replied:
I will check again, but I think it was related to a crash that is not related to this patch

nikolaszimmermann replied:
Will move this to another PR - to fix the case when setting WEBKIT_SKIA_GPU_PAINTING_THREADS to 0.

carlosgcampos commented on Source/WebCore/platform/graphics/texmap/BitmapTexture.cpp:
```diff
@@ -85,33 +85,76 @@ BitmapTexture::BitmapTexture(const IntSize& size, OptionSet<Flags> flags)
+{
+    ASSERT(m_memoryMappedGPUBuffer);
+
+    if (auto eglImage = m_memoryMappedGPUBuffer->createEGLImageFromDMABuf()) {
```
This is always called right after creating the buffer, but it can fail, in which case we don't fallback to use normal texture. Could we make this return a boolean and handle it in BitmapTexture constructor.

nikolaszimmermann replied:
Good catch - I never observed any failure, so I could never tesr these code paths - will adapt.

nikolaszimmermann replied:
Fixed.

carlosgcampos commented on Source/WebCore/platform/graphics/texmap/BitmapTexture.cpp:
```diff
@@ -160,14 +212,38 @@ void BitmapTexture::reset(const IntSize& size, OptionSet<Flags> flags)
+        OptionSet<MemoryMappedGPUBuffer::BufferFlag> bufferFlags;
+        if (flags.contains(Flags::ForceLinearBuffer))
+            bufferFlags.add(MemoryMappedGPUBuffer::BufferFlag::ForceLinear);
+        m_memoryMappedGPUBuffer = MemoryMappedGPUBuffer::create(m_size, bufferFlags);
```
We should null-check the result

carlosgcampos commented on Source/WebCore/platform/graphics/gbm/MemoryMappedGPUBuffer.h:
```diff
@@ -0,0 +1,151 @@
+    bool isMapped() const { return !!m_mappedData; }
+    uint32_t format() const { return m_format; }
+    uint64_t modifier() const { return m_modifier; }
+    bool isLinearBuffer() const;
```
```suggestion
    bool isLinear() const;
```

carlosgcampos commented on Source/WebCore/platform/graphics/texmap/BitmapTexture.h:
```diff
@@ -31,6 +31,9 @@
 #include "IntSize.h"
+#if USE(GBM)
+#include "MemoryMappedGPUBuffer.h"
+#endif
```
Move conditional includes after the main includes block, or just remove the ifdef since the header is already guarded.

carlosgcampos commented on Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedPlatformLayerBufferVideo.cpp:
```diff
@@ -81,8 +81,9 @@ std::unique_ptr<CoordinatedPlatformLayerBuffer> CoordinatedPlatformLayerBufferVi
-    texture->copyFromExternalTexture(textureID);
+    auto size = buffer.size();
+    auto texture = BitmapTexture::create(size);
+    texture->copyFromExternalTexture(textureID, { IntPoint::zero(), size }, { });
```
This also looks unrelated, no?

nikolaszimmermann replied:
See https://github.com/WebKit/WebKit/pull/43411#discussion_r2083634265

carlosgcampos commented on Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedBackingStoreTile.cpp:
```diff
@@ -24,6 +25,9 @@
 #include "BitmapTexture.h"
 #include "CoordinatedTileBuffer.h"
 #include "GraphicsLayer.h"
+#if USE(SKIA)
+#include "SkiaPaintingEngine.h"
+#endif
```
Same here.

carlosgcampos commented on Source/WebCore/platform/mediastream/gstreamer/RealtimeIncomingAudioSourceGStreamer.cpp:
```diff
@@ -24,6 +24,7 @@
 
 #include "GStreamerAudioData.h"
 #include "GStreamerAudioStreamDescription.h"
+#include <wtf/MediaTime.h>
```
Unrelated too

nikolaszimmermann replied:
See https://github.com/WebKit/WebKit/pull/43411#discussion_r2083633796 - I need this to pass the build

carlosgcampos replied:
It can land in a separate patch anyway, as all other unrelated changes

carlosgcampos commented on Source/WebCore/platform/graphics/gbm/MemoryMappedGPUBuffer.cpp:
```diff
@@ -0,0 +1,394 @@
+    return eglImage;
+}
+
+int MemoryMappedGPUBuffer::exportGBMBufferObjectAsDMABuf(unsigned planeIndex)
```
This should return a UnixFileDescriptor to make it clear that it's transferring the ownership.

nikolaszimmermann replied:
Indeed much better.

carlosgcampos commented on Source/WebCore/platform/graphics/gbm/MemoryMappedGPUBuffer.cpp:
```diff
@@ -0,0 +1,394 @@
+void MemoryMappedGPUBuffer::lock()
+{
+    ASSERT(!isLocked());
+    m_isLocked = true;
```
What is this lock? I don't see where this is actually used

nikolaszimmermann replied:
It is used to check that we don't try to obtain an AccessScope multiple times for a single buffer.

nikolaszimmermann replied:
Only relevant for assert-enabled builds though...

carlosgcampos replied:
I find it confusing, tbh, this is not locking/unlocking anything and it's only used on builds with asserts. It would be clearer if it was renamed (maybe beingAccess/endAccess? or something similar) and guarded with `#if ASSERT_ENABLED`

carlosgcampos commented on Source/WebCore/platform/graphics/gbm/MemoryMappedGPUBuffer.h:
```diff
@@ -0,0 +1,154 @@
+    struct gbm_bo* bo() const { return m_bo; }
+    bool isMapped() const { return !!m_mappedData; }
+    uint32_t format() const { return m_format; }
+    uint64_t modifier() const { return m_modifier; }
```
Are these getters really used? I only see isMapped used

nikolaszimmermann replied:
Will check and remove all unncecessary ones -- they were all used at some point, but then surrounding code changed.

carlosgcampos commented on Source/WebCore/platform/graphics/gbm/MemoryMappedGPUBuffer.h:
```diff
@@ -0,0 +1,154 @@
+    OptionSet<BufferFlag> m_flags;
+    bool m_isLocked { false };
+    struct gbm_bo* m_bo { nullptr };
+    unsigned m_planeCount { 0 };
```
Do we need this? can't we just get it from m_bo?

nikolaszimmermann replied:
Ok will adapt.

carlosgcampos commented on Source/WebCore/platform/graphics/gbm/MemoryMappedGPUBuffer.h:
```diff
@@ -0,0 +1,154 @@
+    bool m_isLocked { false };
+    struct gbm_bo* m_bo { nullptr };
+    unsigned m_planeCount { 0 };
+    uint32_t m_format { 0 };
```
Same for the format

carlosgcampos commented on Source/WebCore/platform/graphics/texmap/BitmapTexture.cpp:
```diff
@@ -85,33 +85,76 @@ BitmapTexture::BitmapTexture(const IntSize& size, OptionSet<Flags> flags)
+        // Proceed as usual with GL texture creation if the dma-buf creation failed.
+        // as we only want to allocate the dma-buf, but neither map it, nor create a texture now - but when we
+        // need it (from the thread that needs it!).
+        if (m_memoryMappedGPUBuffer && allocateTextureFromMemoryMappedGPUBuffer())
```
allocateTextureFromMemoryMappedGPUBuffer() already null-checks m_memoryMappedGPUBuffer so we can either remove the check from there or here

carlosgcampos commented on Source/WebCore/platform/graphics/gbm/MemoryMappedGPUBuffer.h:
```diff
@@ -0,0 +1,149 @@
+    uint32_t m_mappedLength { 0 };
+};
+
+inline std::unique_ptr<MemoryMappedGPUBuffer::AccessScope> makeGPUBufferReadScope(MemoryMappedGPUBuffer& buffer)
+{
+    return MemoryMappedGPUBuffer::AccessScope::create(buffer, MemoryMappedGPUBuffer::AccessScope::Mode::Read);
+}
```
Is this used? or do we plan to use this in follow up patches?

nikolaszimmermann replied:
If we want to toy with glReadPixels replacments, this would be needed, so I'd like to keep it.

carlosgcampos commented on Source/WebCore/platform/graphics/gbm/MemoryMappedGPUBuffer.h:
```diff
@@ -0,0 +1,149 @@
+    bool createDMABufFromGBMBufferObject();
+    UnixFileDescriptor exportGBMBufferObjectAsDMABuf(unsigned planeIndex);
+
+    bool isLocked() const { return m_isLocked; }
```
Do we really need a private getter for this private member?

nikolaszimmermann replied:
Removed.

carlosgcampos commented on Source/WebCore/platform/graphics/texmap/BitmapTexture.h:
```diff
@@ -51,6 +55,11 @@ class BitmapTexture final : public ThreadSafeRefCounted<BitmapTexture> {
     enum class Flags : uint8_t {
         SupportsAlpha = 1 << 0,
         DepthBuffer = 1 << 1,
+#if USE(GBM)
+        BackedByDMABuf = 1 << 2,
+        ForceLinearBuffer = 1 << 3,
+#endif
```
Why do we need two flags if they are always set/unset together? I guess force linear buffer implies using dmabuf, no? Why not a single flag DMABufMappeable or something like that?

nikolaszimmermann replied:
No, we can map a super-tiled buffer as well, but then you have to write data in super-tiled format -- I did that in earlier patches.

nikolaszimmermann replied:
Furthermore you can also use `BackedByDMABuf` on the `BitmapTexture` but still use updateContents() to update the pixel data via OpenGL -- so you'll essentially get a GL texture in some proprietary tiled format, but still backed by a dma-buf/EGLImage etc.

carlosgcampos reviewed (approved):
Looks good to me, but I still have a few comments and questions.

nikolaszimmermann committed:
[GTK][WPE] Add support for memory-mapped GPU buffers
https://bugs.webkit.org/show_bug.cgi?id=290800

Reviewed by Carlos Garcia Campos.

When using CPU rendering we produce pixel data that has to be uploaded
into a GL texture for composition - usually using glTexSubImage2D().

In order to save precious memory bandwidth on embedded devices, explore an
alternative mechanism, where buffer uploading happens out of the scope of
OpenGL to avoid overhead, such as copying the pixel data into a staging
buffer in Mesa etc.

The idea is to create a GBM buffer object, export it as a dma-buf, wrap
the dma-buf in an EGLImage and use it as backing for a regular OpenGL
texture, using glEGLImageTargetTexture2DOES(). Later on to update the
texture, one can mmap() the dma-buf fd, to gain write access to the
dma-buf. All memory operations have to be fenced using DMA_BUF_SYNC_*
system calls to ensure that the data is properly synchronized between
CPU and GPU (on shared memory systems, that are common in embedded).

Optimize BitmapTexture::updateContents() for the case when we want
to update a linear, mappable texture (backed by a dma-buf). Avoid
using glCopyTexSubImage2D, and use memcpy() instead on the mapped
memory region -- that avoids driver pitfalls, giving more control
about when synchronization happens.

Allocate tile textures always using GL_NEAREST instead of GL_LINEAR
filter -- GL_NEAREST is less expensive than GL_LINEAR, consuming
less bandwidth. However, it can only be used for textures that are
not scaled/filtered/... to avoid quality loss.

Add WEBKIT_SKIA_USE_LINEAR_TILE_TEXTURES environment variable,
only when set to 1 linear textures are used for the tiles. Off by
default, since the benefit is highly platform dependant.

Canonical link: https://commits.webkit.org/294962@main
